pkg/object/mqttproxy/broker.go

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mqttproxy
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/eclipse/paho.mqtt.golang/packets"
+	"github.com/megaease/easegress/pkg/api"
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/object/function/storage"
+)
+
+type (
+	// Broker is MQTT server, will manage client, topic, session, etc.
+	Broker struct {
+		sync.RWMutex
+		egName string
+		name   string
+		spec   *Spec
+
+		listener net.Listener
+		backend  BackendMQ
+		clients  map[string]*Client
+		auth     map[string]string
+		tlsCfg   *tls.Config
+
+		sessMgr   *SessionManager
+		topicMgr  *TopicManager
+		memberURL func(string, string) ([]string, error)
+
+		// done is the channel for shutdowning this proxy.
+		done chan struct{}
+	}
+
+	// HTTPJsonData is json data received from http endpoint used to send back to clients
+	HTTPJsonData struct {
+		Topic       string `json:"topic"`
+		Qos         int    `json:"qos"`
+		Payload     string `json:"payload"`
+		Base64      bool   `json:"base64"`
+		Distributed bool   `json:"distributed"`
+	}
+)
+
+func newBroker(spec *Spec, store storage.Storage, memberURL func(string, string) ([]string, error)) *Broker {
+	broker := &Broker{
+		egName:    spec.EGName,
+		name:      spec.Name,
+		spec:      spec,
+		backend:   newBackendMQ(spec),
+		clients:   make(map[string]*Client),
+		auth:      make(map[string]string),
+		memberURL: memberURL,
+		done:      make(chan struct{}),
+	}
+	broker.topicMgr = newTopicManager()
+	broker.sessMgr = newSessionManager(broker, store)
+	for _, a := range spec.Auth {
+		broker.auth[a.Username] = a.B64Passwd
+	}
+	err := broker.setListener()
+	if err != nil {
+		logger.Errorf("mqtt.newBroker: broker set listener failed, err:%v", err)
+		return nil
```

Comment:
`newBackendMQ` may create a goroutine, the goroutine leaks in this case.

Reply:
I think the goroutine will not start in this case, since we return before go broker.run()

Reply:
`newBackendMQ` calls `newKafkaMQ`, which creates a goroutine.

Reply:
orz... you are right.